exporter/exporterhelper/internal/bounded_memory_queue.go

```diff
@@ -17,85 +17,144 @@ import (
-		q.stopWG.Add(1)
 		startWG.Add(1)
 		go func() {
 			startWG.Done()
-			defer q.stopWG.Done()
 			for item := range q.items {
-				q.size.Add(^uint32(0))
 				set.Callback(item)
+				q.eventChan <- true
 			}
 		}()
 	}
 	startWG.Wait()
 	return nil
 }
 
-// Produce is used by the producer to submit new item to the queue. Returns false in case of queue overflow.
-func (q *boundedMemoryQueue) Produce(item Request) bool {
-	if q.stopped.Load() {
-		return false
-	}
+func (q *boundedMemoryQueue) eventLoop() {
+	overflow := q.capacity == 0
+	var exitChan chan struct{}
+	for {
+		e := <-q.eventChan
+		if req, ok := e.(Request); ok {
+			if overflow {
+				continue
+			}
+			q.size++
+			q.items <- req
+			overflow = q.capacity == 0 || q.size >= q.capacity
+			if overflow {
+				q.overflow.Store(true)
+			}
+			continue
+		}
+		if done, ok := e.(bool); ok && done {
+			q.size--
+			if exitChan != nil && q.size == 0 {
+				break
+			}
+			previousOverflow := overflow
+			overflow = q.capacity == 0 || q.size >= q.capacity
+			if overflow != previousOverflow {
+				q.overflow.Store(false)
+			}
+			continue
+		}
+		if sizeChan, ok := e.(chan int); ok {
+			sizeChan <- q.size
+			continue
+		}
 
-	// we might have two concurrent backing queues at the moment
-	// their combined size is stored in q.size, and their combined capacity
-	// should match the capacity of the new queue
-	if q.size.Load() >= q.capacity {
-		return false
+		if stopChan, ok := e.(chan struct{}); ok {
+			// mark the event loop stopped.
+			q.stopped.Store(true)
+			// if we have no consumers, empty the queue, dropping its contents.
+			if q.numConsumers == 0 {
+				for len(q.items) > 0 {
+					<-q.items
+					q.size--
+				}
+			}
+			exitChan = stopChan
+			if q.size > 0 {
+				// we have a stop signal, but there are still elements in the queue.
+				continue
+			}
+			break
+		}
 	}
+	close(q.items)
+	close(exitChan)
+	close(q.eventChan)
+}
 
-	q.size.Add(1)
-	select {
-	case q.items <- item:
-		return true
-	default:
-		// should not happen, as overflows should have been captured earlier
-		q.size.Add(^uint32(0))
+// Produce is used by the producer to submit new item to the queue. Returns false in case of queue overflow.
+func (q *boundedMemoryQueue) Produce(item Request) bool {
+	if q.stopped.Load() || q.overflow.Load() {
 		return false
 	}
+	q.eventChan <- item
+	return true
 }
 
 // Stop stops all consumers, as well as the length reporter if started,
 // and releases the items channel. It blocks until all consumers have stopped.
 func (q *boundedMemoryQueue) Stop() {
-	q.stopped.Store(true) // disable producer
-	close(q.items)
-	q.stopWG.Wait()
+	stopChan := make(chan struct{})
+	q.eventChan <- stopChan
+	<-stopChan
 }
 
 // Size returns the current size of the queue
 func (q *boundedMemoryQueue) Size() int {
-	return int(q.size.Load())
+	if q.stopped.Load() {
+		return 0
+	}
+	sizeChan := make(chan int)
+	q.eventChan <- sizeChan
+	return <-sizeChan
```

Comment:
Can I ask why there is a need for a sizeChan? why not take the len of the channel?

Reply:
Because  the size of the queue is something you'd want from asking inside the event loop that controls the queue. If you use a int, you get a data race when you try to read the size. Every other option involves locks.

Reply:
Also, the `q.items` channel doesn't correspond to the size of the queue - that's because the items get removed from `q.items` during consumption but we wait for callbacks.

See:
```
			for item := range q.items {
				set.Callback(item)
				q.eventChan <- true
			}
```

On `set.Callback(item)`, we have a blocking function that the worker blocks on. During that time, the queue is still considered the item as in the queue. All tests use this assumption when looking at queue size.